### ISSUE TITLE
The iat parameter in Trust Mark status requests creates great overhead for no good reason

### ISSUE THREAD

**Issue by Razumain** [openid/federation#24]
The Trust Mark status endpoint takes the issue time as an input parameter.

As the text is written now, the Trust Mark issuer would need to keep track of all issued Trust Marks.

In our implementation, we will issue a new Trust Mark whenever asked, without regard to whether a previous one already exists. There is a very good reason for that. This avoids a complicated algorithm that decides how long a valid TrustMark must be valid in the future in order to deliver it. If we don’t we could ship an existing TrustMark that will expire 1s later, and that is not very useful. By issuing a new one when requested, we make sure that it has the standard lifespan when it is sent.

In order to correctly process the iat parameter, we would have to store issue times for every TrustMark ever issued. And the response will not be very useful if the subject has the requested Trust Mark, but that particular JWT has expired.

In our usage, all the requester is interested in, is if the subject has this TrustMark at this moment in time. No more and no less. We do not plan to store all issued versions of that Trust Mark. It is the verifiers responsibility to check the expiration time. They don’t need the validation service to do that.

I see no legitimate reason for the iat parameter and suggest that it is removed. The status indication is not even assuming the latest issued Trust Mark. This would give a strange result if a status request would be sent for a subject who has that Trust Mark, but where no one has requested a JWT for it in a while. The status response would say false (The latest JWT has expired), but the subject has the Trust Mark, and a new fresh one can be requested.

At least I would request that The Trust Mark issuer is allowed to ignore this parameter if received.

Currently we have to reject all requests that includes iat.

**Comment by Razumain** [comment]
That is part of the problem. There is no discussion about this, but a reader can hardly draw any other conclusion.

If you don't keep track of previously issued Trust Mark, then what are you supposed to do if the "iss" parameter is provided?

The text also says that if "iss" is not provided, then the latest Trust Mark is assumed. That surely also indicate that you need to keep track of issued Trust Marks to know which is last.

I lack a clear description of what a user of this endpoint can expect from a response. Is this a request for the status of a subject in relation to a Trust Mark (i.e. does this subject have this Trust Mark) disregarding any particular issued Trust Mark JWT, or is this a validation service for a particular Trust Mark JWT (Like a X.509 certificate revocation service)?

If this the former (does the subject have this Trust Mark), then "iss" (or "trust_mark") makes no sense. The requester can easily check if the JWT they have has expired, they don't need the validation service to tell them that.

I would personally prefer this to be clarified and the request parameters reduced to just "sub" and "trust_mark_id". Then we could also make this a GET in line with other endpoints and not require a POST.

If I would allow me to be really radical here. I would actually like to delete the whole Trust Mark Status endpoint. For me it is completely redundant and only adds complexity. The difference between returning the status, or just returning a fresh Trust Mark is trivial. If you want to know if a Trust Mark is valid, just ask for a new Trust Mark.

That of course requires the Trust Mark endpoint to be open to all entities. But I really can't se any compelling reason why it shouldn't be that. It's public information spread throughout the federation anyway.

This would materially reduce the complexity of providing and validating Trust Marks. But I assume that such radical changes is a no-go.

**Comment by zachmann** [comment]
In general I agree with @Razumain, however I would not be so radical to remove the trust mark status endpoint all together. 
I would say there's at least a semantic difference between asking for a trust mark status and asking for a trust mark. As you said I would expect the trust mark status endpoint to answer me the following question: "Does this subject have this Trust Mark?" I'm not interested in having the actual trust mark.
Furthermore, it should be cheaper to answer to that question with yes or no than to issue and sign a new trust mark.

Anyway, I would suggest to remove the `iat` parameter. In my opinion it is enough to have the `sub` and `trust_mark_id` parameters and the `trust_mark` parameter could be removed as well. I would argue that the endpoint's purpose is really to answer the question if a certain entity is trust marked with a certain trust mark or not, not if an individual JWT is valid or not; therefore the `trust_mark` would only be used to extract the `sub` and `trust_mark_id`. It might be a shortcut for the client to have the option to request by `trust_mark`, but I think it is not necessary: we can expect the client to extract the information itself from the jwt and the client should already know in which `sub`-`trust_mark_id` combination it is interested.

I can formulate those things and submit a PR if the change is welcomed.

**Comment by zachmann** [comment]
Note that such a trust mark status endpoint would be similar to the trust mark listing endpoint used with a `sub` request parameter, except of the response format.

I would argue that the `sub` request parameter on the listing endpoint is then not needed or the trust mark status endpoint might not be needed.

**Comment by selfissued** [comment]
I'm not sure what you mean by "In order to correctly process the iat parameter, we would have to store issue times for every TrustMark ever issued."  No processing rules are specified for "iat", so I don't know what processing you perform for it.

That said, it's normal for a JWT to indicate when it was issued.  Having that information is a useful signal to recipients for them to use as they choose.  Why should we not provide this signal?

**Comment by zachmann** [comment]
The statement relates to the `iat` request parameter of the trust mark status endpoint: https://openid.github.io/federation/main.html#name-trust-mark-status-request

The spec states:

> OPTIONAL. Number. Time when this Trust Mark was issued. This is expressed as Seconds Since the Epoch, as defined in [[RFC7519](https://openid.github.io/federation/main.html#RFC7519)]. If iat is not specified and the Trust Mark issuer has issued several Trust Marks with the identifier specified in the request to the Entity identified by sub, the most recent one is assumed.

The last sentence implies that if the parameter is present the trust mark issuer should evaluate that specific trust mark; in order to do so it needs to keep track of the individual trust mark jwts.

**Comment by Razumain** [comment]
I agree with all of Gabriels comments here.
I do think that just asking for a Trust Mark would work instead of asking for its status. The cost of doing that is minimal, but I won't insist.

**Comment by rohe** [comment]
In my view if a trust mark issuer issues a new trust mark to an entity it doesn't mean that all previously issued trust marks (of that kind) to that receiver is automatically revoked.

If you in your implementation does that you are perfectly find to do so but there may be good reasons for not revoking previous issued trust marks just because there is a newer.

Yes, this means that a trust mark issuer must keep a list of all trust marks and when they where issued. At least until their expiration time has been reached. 

There are very good reasons for not removing the trust mark status endpoint. So, I will never support that.

I can accept removing the `trust_mark` parameter but not the `iat` one.

I agree with @Razumain that we should not put our trust in TLS. That's one of the pillars of the whole spec.

**Comment by selfissued** [comment]
Fixed by #109

**Comment by zachmann** [comment]
As the iat parameter has been restored, I guess it makes sense to reopen this issue and continue the discussion again. 

Also note Stefan's comment in #25

At a very least it should be allowed for the trust mark issuer to not support the `iat` parameter and response with an `unsupported_parameter` error.

**Comment by selfissued** [comment]
I propose that we address this issue by stating that if a deployment does not have sufficient state to be able to match the `iat` value to a specific trust mark instance, that it MAY ignore the `iat` value and answer the query as if it were not provided.

**Comment by jcmelati** [comment]
> I propose that we address this issue by stating that if a deployment does not have sufficient state to be able to match the `iat` value to a specific trust mark instance, that it MAY ignore the `iat` value and answer the query as if it were not provided.

@selfissued From a consumer POV, not knowing upfront if a parameter that is present on my request will be simply ignored or not. We discussed adding a unique `jti` claim to the issued Trust Mark. If we are proceeding with that, it is more reliable than having to rely on its `iat` time to find a specific instance of a Trust Mark.

That said, the discussion in this thread is much more fundamental than this. It touches the core understanding that I believe we must still reach clarification and agreement on how the life cycle of a Trust Mark instance is expected to look like. We do not share common understanding regarding what the `exp` time in a Trust Mark represents.

From reading the spec, it seems to me that a Trust Mark is akin to a credential, where its `exp` time is directly related to the right derived from the Trust Mark itself. You can add received Trust Marks to your Entity Configuration. The `status` endpoint is then made available so that anyone can check for the status of that specific issued Trust Mark.

If every Trust Mark now is supposed to be short lived, why would anyone add an issued (and soon to be expired) Trust Mark JWT to their Entity Configuration Statement?

**Comment by rohe** [comment]
To add some more complexity to the topic. A trust mark may contain extra information which means that two trust marks of the same kind may not have the same extended content. Add to that, that the extra information may change over time. 

All of this means that you, in the general case, CAN NOT verify a specific trust mark by checking if the issuer can issue a new one of the same type.

@jcmelati This is my take on the life cycle of a Trust Mark. 

- Unless a Trust Mark is revoked it is valid until the **exp** time
- An entity can verify that a Trust Mark is still valid by querying the Trust Mark Status endpoint. The query must contain enough information to identify a specific trust mark instance. Either by providing the complete trust mark instance or by providing enough information (trust mark claims) to identify a specific instance. If the answer to the query is _False_ then the trust mark instance should be regarded as invalid.

In my view there should be no rules or even recommendation about the lifetime of trust marks.

**Comment by selfissued** [comment]
I've come full circle on this, based on considering the implications of this remark of Roland's https://github.com/openid/federation/issues/212#issuecomment-2880642325 :
> Since trust marks can vary over time the trust status endpoint MUST be given enough information to uniquely identify a specific instance of a trust mark to be able to state whether it is valid or not. Hence we need to provide it with the trust_mark_id (or whatever it is called today) and the iat claims or with the complete instance.

The simplest way to uniquely identify a specific instance of a Trust Mark is to pass it by value.  We could do this until draft 40 removed the `trust_mark` parameter to the Trust Mark Status endpoint.  Furthermore, if we remove the `iat` parameter (again), then the implementation need not retain the state needed to map between `iat` values and trust mark instances.  And if we're always passing the whole Trust Mark, then the `trust_mark_id` parameter is unnecessary, since it's available as a claim in the Trust Mark itself.

I believe that passing the Trust Mark by value would resolve all the complexities of and ambiguities about Trust Mark Status that we've been discussing for a long time, including in this issue, #25, #212, and in person during the Federation interop at SUNET.

**Comment by MichaelFraser1999** [comment]
I agree the ability to pass the trust mark JWT in it's entirety would solve a lot of the highlighted issues. +1 to that idea

I would question if removing the trust_mark_id parameter is a mistake though - right now a third part can make a request to establish if a given entity has a given trust mark identifier without needing to store that JWT itself. I highlighted a usecase for this in my third paragraph here: https://github.com/openid/federation/issues/212#issuecomment-2878011697 

Specifically:

> The reason is that a very common pattern in open finance ecosystems is to utilise regulatory roles currently represented as a claim in a signed assertion to determine behaviour. The banks in these systems check if a given RP still has that role when it performs any action. These roles will be represented as trust marks in the new world, so the bank, as an external third party, needs this lookup by trust_mark_id at the issuing party to establish if a client still has this role. This may be performed at a shorter cadence than the client's registration lifetime.

I understand from @rohe's comment that this endpoint is specifically designed to check the status of a JWT instance, however, given the current text in the draft does allow for the above outlined flow (intentional or not) is this functionality we wish to remove? Checking the validity of "would an entity be issued this mark were they asked" is a valid flow imo. Possibly retaining the `trust_mark_id` as an optional parameter to enable the above flow?

**Comment by zachmann** [comment]
> I've come full circle on this, [...]
> The simplest way to uniquely identify a specific instance of a Trust Mark is to pass it by value. 


Yes, I agree and came to the same conclusion as I tried to phrase in #212, in particular

> I personally think we should completely rework it [the trust mark status endpoint] from what it currently is.
From the discussions in https://github.com/openid/federation/issues/24 we removed the `trust_mark` parameter also due to my input. Coming back to the actual and also intuitive meaning of a status endpoint I would argue, that the `trust_mark` parameter should be the only parameter for this endpoint. I now believe the purpose of the trust mark status endpoint is to check if a specific instance of a Trust Mark, i.e. a specific trust mark JWT is still valid or not.

Therefore, I also think that we do not need to preserve the `trust_mark_id` (and `sub`) parameter, since you always will have the JWT if you are interested in a specific JWT.

> If clients are not interested in the question if a certain JWT is valid, but if a certain entity is entitled / eligible to obtain a certain trust mark (id/type/class/label) then it should go to another endpoint.

In particular:
If you need a signed response (there seems to be consensus that this should be the case) about whether a certain entity (sub) has a certain trust mark (trust_mark_id) and maybe also with additional properties (as per rolands comment), I would argue (agreeing with Stefan) that such a statement is basically a trust mark (JWT). Therefore, the straightforward way to fullfil such a request would be to go the trust mark (issuance) endpoint.

From the current state of the draft it would need additional information about those two scenarios and what endpoint would be feasible to use.

**Comment by zachmann** [comment]
What would not be covered by this is what @MichaelFraser1999 mentionend in #212 
> On keeping some form of issuance query parameter on the endpoint, for the same reason mentioned above, its important to be able to distinguish between an entity that had a role, lost it, gained it again. I'd still consider scrapping iat as a filter. Given the scenario outlined, a timestamp parameter may well be more useful? This turns the request into "at this given time, did the specified sub have this trust mark" giving this endpoint historical querying value too

I would argue against this. Such a historical querying capability generally is not supported in the spec - I mean other endpoint also do not support this; I think one could argue in a similar fashion in regards to entities that join a federation, are kicked out, and added again. But the endpoints int he spec only represent the present state; there is @peppelinux draft on [subordinate events](https://github.com/peppelinux/federation-subordinate-events) and I think that might be extendable to also include trust mark issuance events.

**Comment by MichaelFraser1999** [comment]
With the JWT presentation approach, the scenario I mentioned quoted by @zachmann is resolved anyway since that can be infered from any claims inside the JWT. In the ecosystems that would utilise this we can simply use a role_issued_at (or whatever) claim to provide that behaviour so that I'm not too worried about. 

It also re-occurs to me that the scenario I mention here: https://github.com/openid/federation/issues/24#issuecomment-2889388994 could be achieved through the use of the trust mark entities listing endpoint & the sub parameter.
